File: sem5/CN/ham.c
#include <stdio.h>

void main() 
{
    // take data size
    int dataBits;
    printf("Enter number of data bits: ");
    scanf("%d", &dataBits);

    int data[dataBits];
    printf("Enter data bits: ");
    for (int i = 0; i < dataBits; i++) {
        scanf("%d", &data[i]);
    }
    printf("\n");
    for (int i = 0; i < dataBits; i++) {
        printf("%d", data[i]);
    }
    printf("\n\n");

    // make the stuff
    // 2^r >= d + r + 1
    int r = 1;

    while ((1 << r) < dataBits + r + 1) {
        r++;
    }
    printf("Number of redundant bits: %d\n", r);

    int totalBits = dataBits + r;
    int codeword[totalBits + 1]; //1 based indexing cuz otherwise har jageh 1 kon kre

    int dataIdx = dataBits - 1;

    for (int i = totalBits; i >= 1; i--) {
        if ((i & (i - 1)) != 0) {
            if (dataIdx >= 0) {
                codeword[i] = data[dataIdx];
                dataIdx--;
            }
        } else {
            codeword[i] = -1;
        }
    }

    int p_pos = 1;
    for (int i = 0; i < r; i++) {
        int count = 0;

        for (int j = 1; j <= totalBits; j++) {
            if (((j & p_pos) != 0) && (codeword[j] == 1)) {
                count++;
            }
        }
        codeword[p_pos] = count % 2;
        p_pos = p_pos << 1;
    }

    printf("The final codeword is (MSB first): ");
    for (int i = 1; i <= totalBits; i++) {
        printf("%d", codeword[i]);
    }
    printf("\n");

    // take error packet with 1 bit error
    // correct the one bit error
    int error1bit[totalBits + 1];
    printf("Enter bits with 1 bit error: ");
    for (int i = 1; i <= totalBits; i++) {
        scanf("%d", &error1bit[i]);
    }

    int errorPos = 0;
    p_pos = 1;
    for (int i = 0; i < r; i++) {
        int count = 0;
        for (int j = 1; j <= totalBits; j++) {
            if (((j & p_pos) != 0) && (error1bit[j] == 1)){
                count++;
            }
        }
        if (count % 2 != 0) {
            errorPos += p_pos;
        }

        p_pos = p_pos << 1;
    }

    if (errorPos == 0) {
        printf("No error detected.");
    } else {
        error1bit[errorPos] = (error1bit[errorPos] == 0) ? 1 : 0;

        printf("The corrected codeword is: ");
        for (int i = 1; i <= totalBits; i++) {
            printf("%d", error1bit[i]);
        }
        printf("\n");
    }

    // detect 2 bit error
}

File: sem5/CN/ham.py
# Experiment 2: Hamming Code

def calculate_parity(arr, p_bit):
    """Calculates the parity for a given position."""
    count = 0
    for i in range(len(arr)):
        # Check if the bit position should be checked by this parity bit
        if (i + 1) & p_bit:
            if arr[i] == '1':
                count += 1
    return '1' if count % 2 != 0 else '0'

def hamming_encode(data):
    """Encodes data using Hamming Code."""
    n = len(data)
    r = 0
    # Calculate number of redundant bits
    while (2**r < n + r + 1):
        r += 1

    # Create array for codeword, leaving space for parity bits
    codeword = ['0'] * (n + r)
    j = 0
    for i in range(n + r):
        if (i + 1) & i != 0:  # Not a power of 2
            codeword[i] = data[j]
            j += 1

    # Calculate parity bits
    for i in range(r):
        p_pos = 2**i
        codeword[p_pos - 1] = calculate_parity(codeword, p_pos)

    return "".join(codeword)

def hamming_decode_correct(received_code):
    """Detects and corrects a single-bit error in a Hamming code."""
    r = 0
    l = len(received_code)
    while 2**r < l + 1:
        r += 1

    error_pos = 0
    for i in range(r):
        p_pos = 2**i
        parity_val = calculate_parity(list(received_code), p_pos)
        if parity_val == '1':
            error_pos += p_pos

    if error_pos == 0:
        print("No error detected.")
        return received_code
    else:
        print(f"Error detected at position: {error_pos}")
        # Correct the bit
        code_list = list(received_code)
        correct_bit = '0' if code_list[error_pos - 1] == '1' else '1'
        code_list[error_pos - 1] = correct_bit
        return "".join(code_list)

# --- Main Program ---
data_bits = "101"
print(f"Original Data: {data_bits}")

encoded_data = hamming_encode(data_bits)
print(f"Encoded Data (Hamming Code): {encoded_data}")

# Simulate an error at position 5
error_data = list(encoded_data)
error_data[4] = '0' if error_data[4] == '1' else '1'
error_data = "".join(error_data)
print(f"Received Data with error:    {error_data}")

corrected_data = hamming_decode_correct(error_data)
print(f"Corrected Data:              {corrected_data}")